### DISCUSSION TITLE
Issue with Segmentation Fault Using Proxy Driver in PoCL with clblast/clblas Functions  Question

### DISCUSSION THREAD

**Discussion by He32100** [pocl/pocl#1752] [Q&A]
I am using proxy driver to move the Mali GPU to the PoCL platform. The cmake command I used is:
`sudo cmake -DENABLE_ICD=0 -DENABLE_LOADABLE_DRIVERS=0 -DENABLE_PROXY_DEVICE=1 -DWITH_LLVM_CONFIG=/usr/bin/llvm-config-17 ..`
And I have set the environment variable:`export POCL_DEVICES="proxy pthread"`
Now, I can find both the CPU and GPU devices in PoCL platform. The output is as follows:
```
rock@rockpi-4b:~/pocl-55b7a4e6c51b87f29fd48d701670885556f8cb7f/build_proxy$ POCL_BUILDING=1 POCL_DEVICES="proxy cpu" ./bin/poclcc -l
arm_release_ver of this libmali is 'r18p0-01rel0', rk_so_ver is '5'.
LIST OF DEVICES:
0:
  Vendor:   ARM
  Name:     cpu-cortex-a72-cortex-a53
  Version:  OpenCL 3.0 PoCL HSTR: cpu-aarch64-unknown-linux-gnu-cortex-a72
1:
  Vendor:   ARM
  Name:     Mali-T860
  Version:  OpenCL 1.2 v1.r18p0-01rel0.db8fd8841edf74cef96ef24ce665edac HSTR: pocl-proxy: Mali-T860
```
Before including any `CL/*.h` header files, I added `#include "rename_opencl.h".` I want to run an OpenCL program that calls the CLBlast library function `CLBlastSgemm` or the clBLAS library function` clblasSgemm `to perform computations and test if the devices on the PoCL platform are usable. However, regardless of whether I use the CPU or GPU, the program crashes with a segmentation fault.

On the other hand, if I don’t use PoCL and instead use OpenCL, the program runs normally. Additionally, I recompiled PoCL using the cmake command:`cmake -DWITH_LLVM_CONFIG=/usr/bin/llvm-config-17 ..`
And when I use the CPU device from PoCL, the program works fine.
The issue occurs when I build PoCL with the proxy driver and use the clblast or clblas library functions in the program, causing a segmentation fault. Does this problem have anything to do with using the proxy driver to build PoCL?

Here is my OpenCL program:
```
#include </home/rock/pocl-55b7a4e6c51b87f29fd48d701670885556f8cb7f/include/rename_opencl.h>
#include <CL/cl.h>
#include <clblast_c.h>
#include <iostream>
#include <vector>
#include <unistd.h>

// Helper function to check OpenCL errors
#define CHECK_CL_ERROR(error)                                                                    \
    if (error != CL_SUCCESS)                                                                     \
    {                                                                                            \
        std::cerr << "OpenCL error with code " << error << " at line " << __LINE__ << std::endl; \
        exit(EXIT_FAILURE);                                                                      \
    }

// Helper function to check CLBlast errors
#define CHECK_CLBLAST_ERROR(status)                                                                \
    if (status != CLBlastSuccess)                                                                  \
    {                                                                                              \
        std::cerr << "CLBlast error with code " << status << " at line " << __LINE__ << std::endl; \
        exit(EXIT_FAILURE);                                                                        \
    }

int main()
{
    struct timespec start, end;
    double clBlastTime = 0;
    // Initialize OpenCL environment
    cl_platform_id platform;
    cl_device_id device;
    cl_context context;
    cl_command_queue queue;
    cl_int err;

    // Get platform
    err = clGetPlatformIDs(1, &platform, nullptr);
    CHECK_CL_ERROR(err);

    // Get device
    err = clGetDeviceIDs(platform, CL_DEVICE_TYPE_GPU, 1, &device, nullptr);
    CHECK_CL_ERROR(err);

    // Create context
    context = clCreateContext(nullptr, 1, &device, nullptr, nullptr, &err);
    CHECK_CL_ERROR(err);

    // Create command queue
    queue = clCreateCommandQueue(context, device, 0, &err);
    CHECK_CL_ERROR(err);

    // Define matrix dimensions
    const size_t batch_size = 4; // Number of batches
    const size_t m = 16;         // Number of rows in matrix A
    const size_t n = 32;         // Number of columns in matrix B
    const size_t k = 27;         // Number of columns in matrix A and rows in matrix B

    // Create and initialize host-side matrix data
    std::vector<float> host_A(batch_size * m * k, 1.0f); // Each matrix A is initialized to all 1s
    std::vector<float> host_B(batch_size * k * n, 1.0f); // Each matrix B is initialized to all 1s
    std::vector<float> host_C(batch_size * m * n, 0.0f); // Output matrix C is initialized to all 0s

    // Create OpenCL buffers
    cl_mem buffer_A = clCreateBuffer(context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR,
                                     sizeof(float) * batch_size * m * k, host_A.data(), &err);
    CHECK_CL_ERROR(err);

    cl_mem buffer_B = clCreateBuffer(context, CL_MEM_READ_ONLY | CL_MEM_COPY_HOST_PTR,
                                     sizeof(float) * batch_size * k * n, host_B.data(), &err);
    CHECK_CL_ERROR(err);

    cl_mem buffer_C = clCreateBuffer(context, CL_MEM_WRITE_ONLY,
                                     sizeof(float) * batch_size * m * n, nullptr, &err);
    CHECK_CL_ERROR(err);

    // Define offsets for each batch
    std::vector<size_t> a_offsets(batch_size), b_offsets(batch_size), c_offsets(batch_size);
    for (size_t i = 0; i < batch_size; ++i)
    {
        a_offsets[i] = i * m * k;
        b_offsets[i] = i * k * n;
        c_offsets[i] = i * m * n;
    }

    // alpha and beta can be different for each batch, but here we set them both to 1 and 0 for simplicity
    std::vector<float> alphas(batch_size, 1.0f);
    std::vector<float> betas(batch_size, 0.0f);
    CLBlastStatusCode status;

    status = CLBlastSgemm(
        CLBlastLayoutRowMajor,     // Matrices stored in row-major order
        CLBlastTransposeNo,        // Matrix A is not transposed
        CLBlastTransposeNo,        // Matrix B is not transposed
        m, n, k,                   // Dimensions of matrix A (m x k) and matrix B (k x n)
        alphas[0],                 // Alpha value
        buffer_A, a_offsets[0], k, // OpenCL buffer for matrix A, offset, leading dimension of A
        buffer_B, b_offsets[0], n, // OpenCL buffer for matrix B, offset, leading dimension of B
        betas[0],                  // Beta value
        buffer_C, c_offsets[0], n, // OpenCL buffer for matrix C, offset, leading dimension of C
        &queue,                    // Command queue
        nullptr                    // No events needed
    );

    clock_gettime(CLOCK_MONOTONIC, &start);
    // Call CLBlastSgemm to perform batch matrix multiplication
    for (int t = 0; t < 5; ++t)
    {
        for (int i = 0; i < batch_size; ++i)
            status = CLBlastSgemm(
                CLBlastLayoutRowMajor,     // Matrices stored in row-major order
                CLBlastTransposeNo,        // Matrix A is not transposed
                CLBlastTransposeNo,        // Matrix B is not transposed
                m, n, k,                   // Dimensions of matrix A (m x k) and matrix B (k x n)
                alphas[i],                 // Alpha value
                buffer_A, a_offsets[i], k, // OpenCL buffer for matrix A, offset, leading dimension of A
                buffer_B, b_offsets[i], n, // OpenCL buffer for matrix B, offset, leading dimension of B
                betas[i],                  // Beta value
                buffer_C, c_offsets[i], n, // OpenCL buffer for matrix C, offset, leading dimension of C
                &queue,                    // Command queue
                nullptr                    // No events needed
            );
    }
    // Check for CLBlast errors
    CHECK_CLBLAST_ERROR(status);
    // Wait for the queue to finish
    clFinish(queue);

    // Read the results
    err = clEnqueueReadBuffer(queue, buffer_C, CL_TRUE, 0, sizeof(float) * batch_size * m * n, host_C.data(), 0, nullptr, nullptr);
    CHECK_CL_ERROR(err);
    clock_gettime(CLOCK_MONOTONIC, &end);
    clBlastTime = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec) / 1e9;

    // Print the result matrix C (only the first batch)
    std::cout << "ClBlast time: " << clBlastTime << ". Result matrix C (first batch):" << std::endl;

    clReleaseMemObject(buffer_A);
    clReleaseMemObject(buffer_B);
    clReleaseMemObject(buffer_C);
    clReleaseCommandQueue(queue);
    clReleaseContext(context);

    return 0;
}


```

**Comment by linehill** [comment]
Do the libraries link to libpocl instead of libOpenCL and do they have the required `#include "rename_opencl.h"` treatment?

I'm guessing that they are linked to libOpenCL and crash when it receives a PoCL object - not a valid object created by the libOpenCL itself.

**Comment by RABijl** [comment]
The proxy device and ARM has always been a rough combination since their opencl implementation does not return metadata from clGetKernelArgInfo (even when compiling kernels with `-cl-kernel-arg-info`), which the proxy device relies on. 
The proxy is still able to work if a kernel is built for the proxy and another device(CPU in your case), but I don't know if clBLAS builts it for all devices.

But as for your segfault problem, my hunch is that the proxy is somehow incorrectly assuming that the underlying opencl library is providing metadata ([done in the code here](https://github.com/pocl/pocl/blob/672d907a8078c0e29ebf558eae837206fa38158a/lib/CL/devices/proxy/pocl_proxy.cpp#L256)) and is trying to get metadata. A way to confirm this would be to check what `provides_metadata` is set to. If it is set to 1, then that could be why a segfault happens. The reason why a segfault then occurs goes as follows: since ARM is not giving any argument sizes but the proxy assumes it does, at some point the proxy will try to guess it by trying arguments of different sizes. If it does this on a mem_obj argument by providing a pointer to an array, the ARM opencl library segfaults.

**Comment by He32100** [comment]
> Do the libraries link to libpocl instead of libOpenCL and do they have the required `#include "rename_opencl.h"` treatment?
> 
> I'm guessing that they are linked to libOpenCL and crash when it receives a PoCL object - not a valid object created by the libOpenCL itself.



> The proxy device and ARM has always been a rough combination since their opencl implementation does not return metadata from clGetKernelArgInfo (even when compiling kernels with `-cl-kernel-arg-info`), which the proxy device relies on. The proxy is still able to work if a kernel is built for the proxy and another device(CPU in your case), but I don't know if clBLAS builts it for all devices.
> 
> But as for your segfault problem, my hunch is that the proxy is somehow incorrectly assuming that the underlying opencl library is providing metadata ([done in the code here](https://github.com/pocl/pocl/blob/672d907a8078c0e29ebf558eae837206fa38158a/lib/CL/devices/proxy/pocl_proxy.cpp#L256)) and is trying to get metadata. A way to confirm this would be to check what `provides_metadata` is set to. If it is set to 1, then that could be why a segfault happens. The reason why a segfault then occurs goes as follows: since ARM is not giving any argument sizes but the proxy assumes it does, at some point the proxy will try to guess it by trying arguments of different sizes. If it does this on a mem_obj argument by providing a pointer to an array, the ARM opencl library segfaults.

Thank you for the feedback! I've checked the `provides_metadata` setting, and it is indeed set to `0`. Here is the relevant code in lib/CL/devices/proxy/pocl_proxy.c:
```
#ifdef __ANDROID__
    platforms[i].provides_metadata = 0;
#else
    if (strstr(platform_vendor, "ARM"))
        platforms[i].provides_metadata = 0;
    else
        platforms[i].provides_metadata = 1;
#endif
platforms[i].supports_binaries = platforms[i].provides_metadata;
```
Additionally, I tested the program in different configurations:

1. When I don't use PoCL and link the program to`libOpenCL,` it runs fine.
2. After recompiling PoCL using the following cmake command:
`cmake -DWITH_LLVM_CONFIG=/usr/bin/llvm-config-17 ..`
and using the CPU device from PoCL, the program runs without issues.
3. However, when I link the program to `libpocl` and add `#include "rename_opencl.h"`, the program crashes with a segmentation fault, whether I use the CPU or the GPU.

It seems that the issue only occurs when linking to libpocl with the proxy driver. Could this still be related to the proxy driver’s handling of metadata or some other aspect of the PoCL setup?

**Comment by RABijl** [comment]
Oke good that there is indeed not metadata being assumed. It sounds like there might indeed be a mixing up of pocl and other opencl libraries. What you can try is building pocl with: [Proxy_use_libopencl_stub]( https://github.com/pocl/pocl/blob/672d907a8078c0e29ebf558eae837206fa38158a/CMakeLists.txt#L1755). This will allow you to not use rename_pocl and not link to opencl. During runtime it will seach for an opencl library and dlopen it. You can give it hints with the   LIBOPENCL_SO_PATH environment variable.

**Comment by He32100** [comment]
> Oke good that there is indeed not metadata being assumed. It sounds like there might indeed be a mixing up of pocl and other opencl libraries. What you can try is building pocl with: [Proxy_use_libopencl_stub](https://github.com/pocl/pocl/blob/672d907a8078c0e29ebf558eae837206fa38158a/CMakeLists.txt#L1755). This will allow you to not use rename_pocl and not link to opencl. During runtime it will seach for an opencl library and dlopen it. You can give it hints with the LIBOPENCL_SO_PATH environment variable.

Thank you for the suggestion! I followed your advice and built PoCL with libopencl-stub. Here’s how I did it:

1. Built`libopencl-stub`:
```
mkdir build
cd build
cmake ..
make
```

2. Recompiled PoCL using the following command:

`cmake -DPROXY_USE_LIBOPENCL_STUB=1 -DENABLE_ICD=0 -DENABLE_LOADABLE_DRIVERS=0 -DENABLE_PROXY_DEVICE=1 -DWITH_LLVM_CONFIG=/usr/bin/llvm-config-17 -DENABLE_HOST_CPU_DEVICES_OPENMP=0 ..`

3. Set the environment variable:

`export POCL_DEVICES="proxy pthread"`
However, after completing these steps, I still couldn’t detect the proxy device. Below is the output I got:
```
rock@rockpi-4b:~$ export POCL_DEVICES="proxy pthread"
rock@rockpi-4b:~$ clinfo -l
arm_release_ver of this libmali is 'r18p0-01rel0', rk_so_ver is '5'.
Platform #0: ARM Platform
 `-- Device #0: Mali-T860
rock@rockpi-4b:~$ poclcc -l
LIST OF DEVICES:
0:
  Vendor:   ARM
    Name:   cpu-cortex-a72-cortex-a53
 Version:   OpenCL 3.0 PoCL HSTR: cpu-aarch64-unknown-linux-gnu-cortex-a72

```
It seems that the proxy device is still not being detected. Could you provide guidance on how to set the `LIBOPENCL_SO_PATH` environment variable or any additional steps I should take to ensure that both the CPU and GPU proxy devices are available under the PoCL platform?

Thanks again for your help!

**Comment by RABijl** [comment]
The libopencl stub library is integrated into pocl, so you don't have to build it separately. There is a set of usual places it will look for libraries, and the original repo has some pointers on using the evn variables. 

I'm not sure why poclcc is not showing the proxy. iirc, by default it will always show a CPU dev, so maybe check that it is using the variable. You can also set POCL_DEBUG=proxy to see if there is any life there. Also, when building the proxy, the library is named libpocl, you check that poclcc linked to that.

**Comment by He32100** [comment]
Thank you for the suggestions. I’ve followed your advice and set `POCL_DEBUG=proxy`, and upon running poclcc -l, the log shows that PoCL successfully detects multiple available devices, but it still only lists the CPU device:
```
rock@rockpi-4b:~$ poclcc -l
[2025-01-03 02:46:17.546445526067]POCL: in fn pocl_proxy_probe at line 183:
  |     PROXY |  Requested 1 proxy devices.
[2025-01-03 02:46:17.546445868489]POCL: in fn pocl_proxy_probe at line 195:
  |     PROXY |  clGetPlatformIDs found 1 platforms.
[2025-01-03 02:46:17.546446052825]POCL: in fn pocl_proxy_probe at line 265:
  |     PROXY |  Platform 0: The pocl project Devices: 127 OpenGL/EGL interop: 0 Metadata: 1
[2025-01-03 02:46:17.8297073116807077896]POCL: in fn pocl_proxy_probe at line 279:
  |     PROXY |  Requested 1 proxy devices, but more are available (127).
LIST OF DEVICES:
0:
  Vendor:   ARM
    Name:   cpu-cortex-a72-cortex-a53
 Version:   OpenCL 3.0 PoCL HSTR: cpu-aarch64-unknown-linux-gnu-cortex-a72
```
When I set `POCL_DEBUG=all`, I encounter errors indicating `CL_INVALID_DEVICE`:
```
rock@rockpi-4b:~$ export POCL_DEBUG=all
rock@rockpi-4b:~$ poclcc -l
[2025-01-03 02:47:40.545995055409]POCL: in fn pocl_proxy_probe at line 183:
  |     PROXY |  Requested 1 proxy devices.
[2025-01-03 02:47:40.545995271246]POCL: in fn pocl_proxy_probe at line 195:
  |     PROXY |  clGetPlatformIDs found 1 platforms.
[2025-01-03 02:47:40.545995389081]POCL: in fn pocl_proxy_probe at line 265:
  |     PROXY |  Platform 0: The pocl project Devices: 127 OpenGL/EGL interop: 0 Metadata: 1
[2025-01-03 02:47:40.8297073116356339485]POCL: in fn pocl_proxy_probe at line 279:
  |     PROXY |  Requested 1 proxy devices, but more are available (127).
[2025-01-03 02:47:40.545999098842]POCL: in fn POclGetDeviceInfo at line 54:
  |     ERROR | CL_INVALID_DEVICE (!IS_CL_OBJECT_VALID (device))
[2025-01-03 02:47:40.545999161260]POCL: in fn POclGetDeviceInfo at line 54:
......
  |     ERROR | CL_INVALID_DEVICE (!IS_CL_OBJECT_VALID (device))
[2025-01-03 02:47:40.546002786436]POCL: in fn POclGetDeviceInfo at line 54:
  |     ERROR | CL_INVALID_DEVICE (!IS_CL_OBJECT_VALID (device))
[2025-01-03 02:47:40.546005005760]POCL: in fn POclCreateCommandQueue at line 103:
  |   GENERAL |  Created Command Queue 4 (0x7f74018d30) on device 0
[2025-01-03 02:47:40.546005098219]POCL: in fn void pocl_llvm_create_context(cl_context) at line 587:
  |      LLVM |  Created context 3 (0x7f740180e0)
[2025-01-03 02:47:40.544294211]POCL: in fn POclCreateContext at line 232:
  |   GENERAL |  Created Context 3 (0x7f740180e0)
LIST OF DEVICES:
0:
  Vendor:   ARM
    Name:   cpu-cortex-a72-cortex-a53
 Version:   OpenCL 3.0 PoCL HSTR: cpu-aarch64-unknown-linux-gnu-cortex-a72
[2025-01-03 02:47:40.546005243180]POCL: in fn POclReleaseContext at line 53:
  | REFCOUNTS |  Release Context 3 (0x7f740180e0), Refcount: 0
[2025-01-03 02:47:40.8297073116366234126]POCL: in fn POclReleaseContext at line 63:
  | REFCOUNTS |  Free Context 3 (0x7f740180e0)
[2025-01-03 02:47:40.-549211298007]POCL: in fn POclReleaseCommandQueue at line 40:
  | REFCOUNTS |  Release Command Queue 4 (0x7f74018d30), Refcount: 0
[2025-01-03 02:47:40.8297073116366335336]POCL: in fn POclReleaseCommandQueue at line 67:
  | REFCOUNTS |  Free Command Queue 4 (0x7f74018d30)
[2025-01-03 02:47:40.544621466]POCL: in fn void pocl_llvm_release_context(cl_context) at line 592:
  |      LLVM |  releasing LLVM context
```
It seems like the proxy device isn't being properly exposed, even though PoCL recognizes more than one device. In the logs, it looks like PoCL is trying to probe for devices, but it fails when fetching information about them.
Could you suggest any additional configuration steps to ensure that the proxy device is correctly listed and used?

**Comment by pjaaskel** [comment]
I don't think it makes sense to list the proxy device in poclcc as we cannot offline compile for it in PoCL anyhow. The best way I can think of to ensure you have the devices built-in without clinfo is to iterate devices in your OpenCL application.

**Comment by RABijl** [comment]
@He32100 , sorry for the late reply, I built pocl myself to see if there is something broken and these were my results:
### cmake vars:
```
-DENABLE_PROXY_DEVICE=1
-DPROXY_USE_LIBOPENCL_STUB=YES
-DENABLE_ICD=0
-DENABLE_LOADABLE_DRIVERS=0
```
### evn vars:
```
POCL_BUILDING=1
POCL_DEBUG=proxy
POCL_DEVICES="cpu proxy"
```
### command:
```
/home/rabijl/Repositories/pocl/cmake-build-debug/bin/poclcc -l
[2025-01-05 07:54:30.411811595] PoCL: in fn pocl_proxy_probe at line 193:
 |     PROXY |  Requested 1 proxy devices.
[2025-01-05 07:54:30.419513019] PoCL: in fn pocl_proxy_probe at line 205:
 |     PROXY |  clGetPlatformIDs found 1 platforms.
[2025-01-05 07:54:30.419526324] PoCL: in fn pocl_proxy_probe at line 279:
 |     PROXY |  Platform 0: NVIDIA Corporation Devices: 1 OpenGL/EGL interop: 0 Metadata: 0
LIST OF DEVICES:
0:
  Vendor:   AuthenticAMD
    Name:   cpu-znver4-AMD Ryzen 9 7900X 12-Core Processor
 Version:   OpenCL 3.0 PoCL HSTR: cpu-x86_64-pc-linux-gnu-znver4
1:
  Vendor:   NVIDIA Corporation
    Name:   NVIDIA GeForce GTX 1060 6GB
 Version:   OpenCL 3.0 CUDA HSTR: pocl-proxy: NVIDIA GeForce GTX 1060 6GB

Process finished with exit code 0
```
In you case, the proxy found pocl somehow instead of the underlying opencl library.  We have done some work with running [pocl on android](https://github.com/cpc/PoCL-R-Reference-Android-Java-Client/blob/main/android/app/src/main/cpp/CMakeLists.txt) which might give you some hints. In this case we build `libpocl.so` which is dlopened by JOCL, but there are a lot of similarities to what you are doing.

**Comment by He32100** [comment]
Thanks for your reply! 
But I still couldn't solve the problem. So I attempted to use PoCL-Remote to move Mali GPU to the PoCL platform. Below are the steps I followed and the issue I encountered:

1. I recompiled PoCL with the following cmake configuration:
```
cmake -DENABLE_HOST_CPU_DEVICES=1 \
      -DENABLE_LLVM=1 \
      -DWITH_LLVM_CONFIG=/usr/bin/llvm-config-17 \
      -DENABLE_LOADABLE_DRIVERS=0 \
      -DENABLE_ICD=1 \
      -DENABLE_REMOTE_CLIENT=1 \
      -DENABLE_REMOTE_SERVER=1 ..
make
make install
sudo cp /usr/local/etc/OpenCL/vendors/pocl.icd /etc/OpenCL/vendors/

```

2. I launched pocld on a separate terminal with the following command:
`pocld -a 192.168.199.23 -p 7777 --log_filter=err,warn,general`
The log output indicates that the server started successfully:
```
[2025-01-06 02:20:54.545717787595]POCL: in fn launch at line 364:
  |   GENERAL |  Server PID=2955022 listening for client connections on rockpi-4b.lan (192.168.199.23), ports 7777 (command), 7778 (stream)
[2025-01-06 02:20:54.545718144892]POCL: in fn listen_peers at line 89:
  |   GENERAL |  PL: listening for peers on port 7779 (tcp)

```

3. On the client side, I set the following environment variables in the build directory:
```
export OCL_ICD_VENDORS=$PWD/ocl-vendors/pocl-tests.icd
export POCL_DEVICES="remote remote"
export POCL_REMOTE0_PARAMETERS=192.168.199.23:7777/0
export POCL_REMOTE1_PARAMETERS=192.168.199.23:7777/1
```

4. However, when I ran the clinfo -l command, no devices were found, and pocld crashed with a segmentation fault. The pocld log output was as follows:
```
rock@rockpi-4b:~$ pocld -a 192.168.199.23 -p 7777 --log_filter=err,warn,general
[2025-01-06 02:20:54.545717787595]POCL: in fn launch at line 364:
  |   GENERAL |  Server PID=2955022 listening for client connections on rockpi-4b.lan (192.168.199.23), ports 7777 (command), 7778 (stream)
[2025-01-06 02:20:54.545718144892]POCL: in fn listen_peers at line 89:
  |   GENERAL |  PL: listening for peers on port 7779 (tcp)
[2025-01-06 02:21:48.546315693493]POCL: in fn readAllClientSocketsThread at line 537:
  |   GENERAL |  Client socket poll returned 1 fds with events
[2025-01-06 02:21:48.688051770874]POCL: in fn operator() at line 577:
  |   GENERAL |  Accepted client command connection from rockpi-4b.lan (192.168.199.23)
[2025-01-06 02:21:48.546317909899]POCL: in fn readAllClientSocketsThread at line 537:
  |   GENERAL |  Client socket poll returned 1 fds with events
[2025-01-06 02:21:48.2941692677601068]POCL: in fn read at line 193:
  |   GENERAL |  -------------------------------------------------------------------------------
[2025-01-06 02:21:48.3545277504050064569]POCL: in fn read at line 195:
  |   GENERAL |  MESSAGE RECEIVED, ID: 0 TYPE: CreateOrAttachSession SIZE: 88/88 
[2025-01-06 02:21:48.3545277504050078569]POCL: in fn read at line 302:
  |   GENERAL |  ALL READS COMPLETE FOR ID: 0, fd=7
[2025-01-06 02:21:48.546318068568]POCL: in fn performSessionSetup at line 458:
  |   GENERAL |  Registered new client session 1 d347bb00493637504ef156694cb906eb
[2025-01-06 02:21:48.365929644555]POCL: in fn TrafficMonitor at line 37:
  |   GENERAL |  POCL_TRAFFIC_LOG_DIR not set, stats will not be written to disk
arm_release_ver of this libmali is 'r18p0-01rel0', rk_so_ver is '5'.
[2025-01-06 02:21:48.365993011495]POCL: in fn SharedCLContext at line 545:
  |   GENERAL |  Platform 0 has 1 devices
Segmentation fault
```
How do I properly use POCl-Remote to move my mali gpu to pocl platform?

**Reply by RABijl** [reply]
If you want to go the route with the remote instead of proxy, I'd make sure pocld only grabs the mali opencl lib and set POCL_DEVICES="cpu remote".

**Reply by He32100** [reply]
I have set POCL_DEVICES="cpu remote", but I still get Segmentation fault.

**Reply by pjaaskel** [reply]
It's difficult to guess what's wrong. Are you able to debug a bit by running in gdb and dumping a back trace?
